File: frontend/src/utils/logger.js
/**
 * Unified Logger for WAKE App Frontend
 *
 * @fileoverview Provides a level-aware logger with the DSC: prefix for all
 * frontend logging. All modules must use this logger instead of raw
 * console.log, console.warn, etc. The active level defaults to DEBUG in
 * development and INFO in production; it can be overridden at runtime via
 * `logger.setLevel(name)` — typically called once after the app fetches
 * `/api/v1/config` so the backend `LOG_LEVEL` env var drives both sides.
 *
 * @author David Seguin
 * @version 1.0.0
 * @since 2026
 * @license Professional - All Rights Reserved
 *
 * Key Features:
 * - Five levels: NONE / ERROR / WARN / INFO / DEBUG
 * - Runtime level override via `setLevel`
 * - Persists chosen level to localStorage so reloads stay consistent
 *
 * Dependencies: none
 *
 * Security Considerations:
 * - Never log sensitive data (tokens, passwords)
 */

const LEVELS = { NONE: 0, ERROR: 1, WARN: 2, INFO: 3, DEBUG: 4 };
const STORAGE_KEY = 'logLevel';

function resolveInitialLevel() {
  const stored = typeof localStorage !== 'undefined' ? localStorage.getItem(STORAGE_KEY) : null;
  if (stored && stored in LEVELS) return LEVELS[stored];
  return import.meta.env.DEV ? LEVELS.DEBUG : LEVELS.INFO;
}

let activeLevel = resolveInitialLevel();

function emit(level, consoleMethod, tag, args) {
  if (activeLevel < level) return;
  consoleMethod(`DSC: [${tag}]`, ...args);
}

const logger = {
  /**
   * Set the active log level. Accepts a level name (e.g. 'INFO') or number.
   * Persisted to localStorage so subsequent loads keep the same level.
   *
   * @param {('NONE'|'ERROR'|'WARN'|'INFO'|'DEBUG'|number)} level
   */
  setLevel(level) {
    const resolved = typeof level === 'number' ? level : LEVELS[String(level).toUpperCase()];
    if (resolved === undefined) return;
    activeLevel = resolved;
    try {
      const name = Object.keys(LEVELS).find((k) => LEVELS[k] === resolved);
      if (name) localStorage.setItem(STORAGE_KEY, name);
    } catch (_err) { /* storage unavailable */ }
  },

  getLevel() {
    return Object.keys(LEVELS).find((k) => LEVELS[k] === activeLevel) || 'INFO';
  },

  debug: (...args) => emit(LEVELS.DEBUG, console.debug, 'DEBUG', args),
  info:  (...args) => emit(LEVELS.INFO,  console.info,  'INFO',  args),
  warn:  (...args) => emit(LEVELS.WARN,  console.warn,  'WARN',  args),
  error: (...args) => emit(LEVELS.ERROR, console.error, 'ERROR', args),
};

export { LEVELS };
export default logger;

File: backend/utils/logger.py
"""
Unified Logger for Base App Backend

@fileoverview Provides a configured logger instance with the DSC: prefix
for all backend logging. All modules must use this logger instead of
print() or direct logging calls.

@author David Seguin
@version 1.0.0
@since 2026
@license Professional - All Rights Reserved

Key Features:
- Consistent DSC: prefix on all log messages
- Structured log format with timestamp, level, and message
- Console output with color-coded levels
- Configurable log level via environment variable

Dependencies:
- Python standard library logging module

Security Considerations:
- Never log sensitive data (tokens, passwords, secrets)
- Sanitize user input before logging

Performance Notes:
- Minimal overhead; standard library logging
"""

import logging
import os


class _RequestIdFilter(logging.Filter):
    """Inject g.request_id (or '-') into every log record."""

    def filter(self, record):  # noqa: A003 - logging API
        request_id = '-'
        try:
            from flask import g, has_request_context
            if has_request_context():
                request_id = getattr(g, 'request_id', '-') or '-'
        except Exception:
            pass
        record.request_id = request_id
        return True


LOG_LEVEL = os.environ.get('LOG_LEVEL', 'DEBUG').upper()

formatter = logging.Formatter(
    fmt='DSC: [%(levelname)s] %(asctime)s [%(request_id)s] - %(message)s',
    datefmt='%Y-%m-%d %H:%M:%S'
)

handler = logging.StreamHandler()
handler.setFormatter(formatter)
handler.addFilter(_RequestIdFilter())

logger = logging.getLogger('base-app')
logger.setLevel(getattr(logging, LOG_LEVEL, logging.DEBUG))
logger.addHandler(handler)
logger.propagate = False
